msal/managed_identity.py

```diff
@@ -0,0 +1,592 @@
+    ID = "Id"
+
+    # Valid values for key ID_TYPE
+    CLIENT_ID = "ClientId"
+    RESOURCE_ID = "ResourceId"
+    OBJECT_ID = "ObjectId"
+    SYSTEM_ASSIGNED = "SystemAssigned"
+
+    _types_mapping = {  # Maps type name in configuration to type name on wire
+        CLIENT_ID: "client_id",
+        RESOURCE_ID: "mi_res_id",
+        OBJECT_ID: "object_id",
+    }
+
+    @classmethod
+    def is_managed_identity(cls, unknown):
+        return isinstance(unknown, ManagedIdentity) or (
+            isinstance(unknown, dict) and cls.ID_TYPE in unknown)
+
+    @classmethod
+    def is_system_assigned(cls, unknown):
+        return isinstance(unknown, SystemAssignedManagedIdentity) or (
+            isinstance(unknown, dict)
+            and unknown.get(cls.ID_TYPE) == cls.SYSTEM_ASSIGNED)
+
+    @classmethod
+    def is_user_assigned(cls, unknown):
+        return isinstance(unknown, UserAssignedManagedIdentity) or (
+            isinstance(unknown, dict)
+            and unknown.get(cls.ID_TYPE) in cls._types_mapping
+            and unknown.get(cls.ID))
+
+    def __init__(self, identifier=None, id_type=None):
+        # Undocumented. Use subclasses instead.
+        super(ManagedIdentity, self).__init__({
+            self.ID_TYPE: id_type,
+            self.ID: identifier,
+        })
+
+
+class SystemAssignedManagedIdentity(ManagedIdentity):
+    """Represent a system-assigned managed identity.
+
+    It is equivalent to a Python dict of::
+
+        {"ManagedIdentityIdType": "SystemAssigned", "Id": None}
+
+    or a JSON blob of::
+
+        {"ManagedIdentityIdType": "SystemAssigned", "Id": null}
+    """
+    def __init__(self):
+        super(SystemAssignedManagedIdentity, self).__init__(id_type=self.SYSTEM_ASSIGNED)
+
+
+class UserAssignedManagedIdentity(ManagedIdentity):
+    """Represent a user-assigned managed identity.
+
+    Depends on the id you provided, the outcome is equivalent to one of the below::
+
+        {"ManagedIdentityIdType": "ClientId", "Id": "foo"}
+        {"ManagedIdentityIdType": "ResourceId", "Id": "foo"}
+        {"ManagedIdentityIdType": "ObjectId", "Id": "foo"}
+    """
+    def __init__(self, *, client_id=None, resource_id=None, object_id=None):
+        if client_id and not resource_id and not object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.CLIENT_ID, identifier=client_id)
+        elif not client_id and resource_id and not object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.RESOURCE_ID, identifier=resource_id)
+        elif not client_id and not resource_id and object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.OBJECT_ID, identifier=object_id)
+        else:
+            raise ManagedIdentityError(
+                "You shall specify one of the three parameters: "
+                "client_id, resource_id, object_id")
```

Comment:
It is a common misunderstanding that **user-assigned managed identity requires an ID**. However, this is not true.

When there is no system-assigned managed identity configured and there is only one user-assigned managed identity, you may omit ID to acquire an access token for that user-assigned managed identity. For more details, see https://github.com/Azure/azure-cli/issues/13188

Reply:
This a behavior that we want to get away from...

Reply:
Unless managed identity service rejects the "default user-assigned managed identity" usage, this pattern causes confusion.

Reply:
@neha-bhargava  to also provide a perspective.

Reply:
Afaik, this is handled by Managed Identity API itself, not by the SDK:

- if the app developer configures a SAMI
- no SAMI exists
- but a single UAMI exists
- then that UAMI is used

Reply:
I understand the user can use `SystemAssignedManagedIdentity` to trigger this behavior, but the name `SystemAssignedManagedIdentity` becomes misleading in that case, as a user-assigned managed identity is actually used.

Reply:
Quoting our offline conclusion.

1. That is the underlying behavior baked into Azure VM's endpoint (but not on any other flavors of Managed Identities). I left [a trace here in MSAL code base](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/480/files#diff-24c0727ff4626c6c535d05c13b61fa4b4a47d6fc4496ec0ceadc734191de19cbR235-R242), which points to [Azure VM's official documentation](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http). We do not like it or advertise it, but, that behavior on the endpoint layer will probably stay forever, until/unless we drop that endpoint and switch to Managed Identity v2 in the future.

2. With our programming model of encouraging/enforcing the choice between SAMI and UAMI, we at least conceptually guide app developers to make that decision upfront. So, when app developer wants to use the only UAMI on their VM, they are expected to create a UAMI with its client_id. They are not supposed to use SAMI. There is no DefaultManagedIdentity concept here. There will be no ambiguity in this usage pattern.
Of course, nothing stops app developer from knowing `#1` and then purposely (ab)use our SAMI interface to get a token that is actually for the UAMI. But we (MSAL) did our part.